test/checkpoint/workers.c

```diff
@@ -239,56 +239,51 @@ real_worker(void)
-                testutil_check(session->commit_transaction(session, NULL));
-                for (j = 0; j < g.ntables; j++)
-                    testutil_check(cursors[j]->close(cursors[j]));
-                has_cursors = false;
-                __wt_readlock((WT_SESSION_IMPL *)session, &g.clock_lock);
-                testutil_check(session->begin_transaction(session, begin_cfg));
-            }
-            testutil_check(__wt_snprintf(buf, sizeof(buf), "commit_timestamp=%x", g.ts + 1));
-            testutil_check(session->timestamp_transaction(session, buf));
-            __wt_readunlock((WT_SESSION_IMPL *)session, &g.clock_lock);
-
-            for (j = 0; !has_cursors && j < g.ntables; j++)
-                if ((ret = session->open_cursor(
-                       session, g.cookies[j].uri, NULL, NULL, &cursors[j])) != 0) {
-                    (void)log_print_err("session.open_cursor", ret, 1);
-                    goto err;
-                }
-            has_cursors = true;
-        }
-        for (j = 0; ret == 0 && j < g.ntables; j++) {
+        for (j = 0; ret == 0 && j < g.ntables; j++)
             ret = worker_op(cursors[j], keyno, i);
-        }
         if (ret != 0 && ret != WT_ROLLBACK) {
             (void)log_print_err("worker op failed", ret, 1);
             goto err;
         } else if (ret == 0 && __wt_random(&rnd) % 7 != 0) {
-            if ((ret = session->commit_transaction(session, NULL)) != 0) {
-                (void)log_print_err("real_worker:commit_transaction", ret, 1);
-                goto err;
+            if (g.use_timestamps) {
+                if (__wt_try_readlock((WT_SESSION_IMPL *)session, &g.clock_lock) == 0) {
+                    testutil_check(
+                      __wt_snprintf(buf, sizeof(buf), "commit_timestamp=%x", g.ts + 1));
+                    __wt_readunlock((WT_SESSION_IMPL *)session, &g.clock_lock);
+                    if ((ret = session->commit_transaction(session, buf)) != 0) {
+                        (void)log_print_err("real_worker:commit_transaction", ret, 1);
+                        goto err;
+                    }
+                    restart_txn = true;
+                }
+            } else {
+                if ((ret = session->commit_transaction(session, NULL)) != 0) {
+                    (void)log_print_err("real_worker:commit_transaction", ret, 1);
+                    goto err;
+                }
+                restart_txn = true;
             }
         } else {
             if ((ret = session->rollback_transaction(session, NULL)) != 0) {
                 (void)log_print_err("real_worker:rollback_transaction", ret, 1);
                 goto err;
             }
+            restart_txn = true;
+        }
+        if (restart_txn) {
+            if ((ret = session->begin_transaction(session, NULL)) != 0) {
+                (void)log_print_err("real_worker:begin_transaction", ret, 1);
+                goto err;
+            }
+            restart_txn = false;
```

Comment:
Could you change this to `start_txn` and have it as the first thing the loop does? Then you wouldn't need the `begin_transaction` before the loop *and* this one at the end.

I'm also curious: does WT_SESSION::close automatically commit or abort a transaction? i.e: after the loop finishes, does it matter whether a transaction is running or not?

Reply:
WT_SESSION::close does rollback any active transaction. So it doesn't matter.